### PR TITLE
Fix Problem on Create Database

### LINKED ISSUE #2588: Error occured while getting a DB profile for 'None'
This error seems like is intermittent, happens from time to time. But not always.
First run:

```
root@<machine>:~/InstaPy# python3 <file_name>.py 
WARNING [2018-08-01 07:53:25] [<username>]  Heeh! Error occured while getting a DB profile for 'None':
	b'NOT NULL constraint failed: profiles.name'
Traceback (most recent call last):
  File "<file_name>.py", line 24, in <module>
    multi_logs=True)
  File "/root/InstaPy/instapy/instapy.py", line 192, in __init__
    get_db(True)
  File "/root/InstaPy/instapy/database_engine.py", line 75, in get_db
    id = get_profile(name, address, logger) if (id is None or make==True) else id
  File "/root/InstaPy/instapy/database_engine.py", line 110, in get_profile
    profile = dict(profile)
TypeError: 'NoneType' object is not iterable
```

Second run:
```
root@<machine>:~/InstaPy# python3 <file_name>.py 

OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
INFO [2018-08-01 07:53:29] [<username>]  Session started!
```

## Expected Behavior
Run python file with actions

## Current Behavior
Intermitent error
```
WARNING [2018-08-01 07:53:25] [<username>]  Heeh! Error occured while getting a DB profile for 'None':
	b'NOT NULL constraint failed: profiles.name'
```

### PR DESCRIPTION
**Pull request by rafaelfcsouza**
Small refactor on Create Database and Get Database Logic on instapy/database_engine.py
Separated logic into small functions for better redability and removed comments (functions names have full name now)
Separated SQL into constants

Fixes #2588

### CODE REVIEW

**Comment by uluQulu** [comment]
The database engine is so simple to be separated where separation makes more complexity here.
I have thought about writing _table making_ optional, but all of the tables are in need for every usage and needs to be started as gets a database.
Also the ultra situation handling is lost with these changes,.  
Etc., etc., all those changes does not fit the smooth design considerations I have offered.  

I don't know other guys, but those are the extra and unneeded changes for me that I can't uppervote.

For that bug- discussed in #2578, I am sure it has a simple solution if any of the affected users give a good feedback.  


Cheers 😁

**Comment by rafaelfcsouza** [comment]
Sorry man, prefer functions over comments.

As stated in many refactoring books, a functions should do one thing only. If you have comments on you function, probably its too long.

If I missed any logic please let me know. Since there aren’t unit tests I may have missed it.

**Comment by samrap** [comment]
@rafaelfcsouza Is correct.

I'm sure I speak for many users that we all appreciate the hard work the maintainers put into this project. Seriously, this thing does some pretty incredible work! However, it is pretty clear when having to dig through the source code can be messy in certain areas. Duplicate logic and functions that are hundreds of lines long. Perhaps that doesn't matter to the maintainers, which is fine, but this PR should be merged. Not only does it fix the issue, but it introduces better maintainability down the line. If the PR missed some features, then we should work to get those fixed. But overall, the refactor has a +1 from me. Thanks for your work @rafaelfcsouza !

**Comment by uluQulu** [comment]
Hey @samrap 
+ "_..it doesn't matter to the maintainers,.._"
It matters a lot to me and I am the one who wrote that utterly simple file for its duties.
+ "_if the PR missed some features.._"
This PR brings no any features, rather, it is a separation of two functions- each `69` and `36` lines of codes into easter eggs to and misses the flexibility and the smoothness of the original design.
 + "_it introduces better maintainability down the line_" 
Added this extra unneeded complexity, it will stuck the future maintainers' mind `down the line`
+ "_Duplicate logic_"
There is no any duplicate logic, show me a one?
+ "_we all appreciate the hard work_"  
Me too.   
+ "_when having to dig through the source code can be messy in certain areas_"
I think it's a sign of the sophisticated code which makes hard for you to understand but there is nothing messy in that file..  
+ "_Not only does it fix the issue_"  
It has been fixed **simply** as I have said, see #2607 for **`1` line of fix**  

----------------

@rafaelfcsouza 
+ "_prefer functions over comments_"
I don't think functions can replace comments as of today 😂 so I will not prefer functions over comments.
+ "_a functions should do one thing only._"
It is a perfect reality which does not exist cos it is not optimal.  

+ "_if you have comments on you function, probably its too long._"
Functions and comments are different things, actually with your understanding, if you separate a line with a comment to a function then in your perfect reality you have to add a comment **again** to the start of that function otherwise people will not understand why it is written -> **extra complexity**.  
_I think if any line of code in anywhere needs a comment, then do add a comment, it is fine, comments are cool_ 😎

-----------------------

The thing is, @rafaelfcsouza, comments are so useful notion in programming and everywhere I read says use comments whenever required.  
If you think so, then make 27 separate functions out of the 27 lines of comments in `click_element` (_total of 48 lines_) definition.  
Or @samrap, if you think every `100` long function must be separated, the have a look at `like_by_tags`- `204` lines of code. 

_I think a code block must be separated into a function if there is a repetition OR if it is gonna make things look much more simpler but not complex._  

---------------

The reason I said these is cos @samrap's comment targeted me and I am the one who designed and implemented this particular codeblock AND it can be that I am **not** a professional programmer, yet, but I do take all of the considerations before writing a line of code.  


Cheers 😁

**Comment by rafaelfcsouza** [comment]
I really recommend you read Clean Code by Robert Martin. You really should.
And I'm sorry if I sound rude, but I'm currently not with that much time to argue about quality of code.

I'll leave here a few links for further reading:
https://medium.com/@amlcurran/clean-code-the-curse-of-a-boolean-parameter-c237a830b7a3
https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/
https://blog.codinghorror.com/coding-without-comments/
https://code.tutsplus.com/tutorials/refactoring-legacy-code-part-2-magic-strings-constants--cms-20527
https://refactoring.com/catalog/extractMethod.html
https://en.wikipedia.org/wiki/Cyclomatic_complexity

Thank you, have a nice day :)
And Thank you for fixing the bug.

**Comment by timgrossmann** [comment]
@rafaelfcsouza @uluQulu I'd agree with Rafael in this case.
The code looks cleaner and is easier to read since the "main" logic is separated out into the distinct functions.

@rafaelfcsouza Could you resolve the conflicts @rafaelfcsouza please?
I merged #2607 which fixes the DB error, however I think this code change helps people getting into the project understanding the code more easily.
I'll review your code once more to make sure that the code still works like the old one, once you've resolved the conflicts.

Thanks guys, this community is just awesome!

P.S. @uluQulu Your work is still incredible and mandatory for this project, I just think making the methods shorter and do a "separation of concerns" is the way to go in order to make it easier for future contributors to touch the code.

**Comment by uluQulu** [comment]
That's all right, @timgrossmann 
Thank you for your support, guidance and encouragement. I am thankful to be a part of this team.  

@rafaelfcsouza seems to be a more experienced programmer than me since I have not finished a programming book to day.  But I will highlight some technical details again.  

+ It is always updating `Settings` db location  == **waste**

+ `create_database` sounds like it is always creating a database but actually it is not, it is just **getting a database**.

+ `1` line of functions which are gonna be used only once == feels so wasted, respect functions!
(_the others are mostly `2` lines and `4` line I saw_)

+ `create_database(self.username)` sounds like creating a database for a username? - nah, the database is unique, it is just adding a profile for a username inside that DB

+ It is like a maze, one can get lost trying to modify something amongst little function**rinhos** 😂  
_Actually these lots of small functions make code harder to understand._  
  
I won't stay here so long, @timgrossmann, I understand you point. 🎉🎉🎉


Cheers 😁

File: instapy/database_engine.py [hunk 1]
```diff
@@ -2,119 +2,138 @@
 import sqlite3
 
 from .settings import Settings
-from .time_util import sleep
 
+FOLLOW_RESTRICTION_TABLE_STATEMENT = """CREATE TABLE IF NOT EXISTS `followRestriction` (
```
Comment (timgrossmann) [inline instapy/database_engine.py]: Please align to other Constants declaration (move to next line)

File: instapy/database_engine.py [hunk 2]
```diff
@@ -2,119 +2,138 @@
...
+    cursor.execute(RECORD_ACTIVITY_TABLE_STATEMENT)
+
+
+def create_table_profiles(cursor):
+    cursor.execute(PROFILES_TABLE_STATEMENT)
+
+
+def create_table_follow_restriction(cursor):
+    cursor.execute(FOLLOW_RESTRICTION_TABLE_STATEMENT)
+    pass
```
Comment (timgrossmann) [inline instapy/database_engine.py]: Why do we need a pass here?

File: instapy/util.py [hunk 3]
```diff
@@ -3,7 +3,6 @@
 import datetime
 import os
 import re
-import random
```
Comment (timgrossmann) [inline instapy/util.py:6]: Unused import?

File: instapy/util.py [hunk 4]
```diff
@@ -12,13 +11,11 @@
 from selenium.common.exceptions import NoSuchElementException
 from selenium.common.exceptions import WebDriverException
 
-from .settings import Settings
```
Comment (timgrossmann) [inline instapy/util.py:15]: Unused import?

Comment (rafaelfcsouza) [reply instapy/util.py:15]: Yes

File: instapy/util.py (hunk 3 above)
Comment (rafaelfcsouza) [reply instapy/util.py:6]: Yes

File: instapy/database_engine.py (hunk 2 above)
Comment (rafaelfcsouza) [reply instapy/database_engine.py]: Generated automatically :( with Refactor > extract from PyCharm

File: instapy/database_engine.py (hunk 1 above)
Comment (rafaelfcsouza) [reply instapy/database_engine.py]: I'm fixing the PR

Comment (timgrossmann) [reply instapy/database_engine.py]: Thank you very much!

**Comment by timgrossmann** [comment]
@uluQulu Thank you for understanding this, I really hope this will not affect future contributions 😉 

You're a very good programmer, but there always is someone with a little bit more experience, especially when it comes to big(ger) projects. I think @rafaelfcsouza is the right guy here!
So thank you very much for making the changes.

@uluQulu I totally agree that it might be weird for people first when they see the structure, but in most cases, they, hopefully, use some code editor which will highlight the associated method so they can find their way through the "maze". 
I totally understand your point here and we definitely need to make sure to not overdue this "separation" but looking at a lot of methods in InstaPy, I should've started refactoring a lot of it when it started to grow, most likely it would've made things easier in this stage.

Just to make sure you got that right, `create_database` only get's called once. When the db actually should be created (at the start of the script, the only position you used `make=True`).
All the other locations where the method was called without True, he uses the `get_db` method which only gets the DB.

I totally agree on the `create_database` name in terms of "it is just adding a profile for a username inside that DB". We might want to rename this to `add_user_to_db` or something like that which makes it clearer that the DB is not "recreated".

Considering the updating the location DB part, I think I don't fully understand.
It seems like he just used most of your code and moved it into separate methods.
Going through it, it looks like the DB is only updated once, just as with your code.
@rafaelfcsouza @uluQulu Am I missing something here?

Again, thank you very much for this discussion and the PR/Code.
I am learning so much from you guys, I'm super happy to have such a community around our project which just makes it better (even though sometimes there are little bugs 😉)

Seriously, thank you
Best
Tim

**Comment by rafaelfcsouza** [comment]
I'm sure that there's still a lot of refactor to be done in this class. But that's what I cloud do.
"Any time someone sees some code that isn't as clear as it should be, they should take the opportunity to fix it"

Hope I helped.

**Comment by timgrossmann** [comment]
@rafaelfcsouza Let's hope it works for everybody as nicely as it does for me 😉

**Comment by uluQulu** [comment]
Thanks a bunch, @timgrossmann.
But I have no activity as before if you noticed and I came back only to write the modules I promised earlier and it seems to be done. 
@rafaelfcsouza is an experienced programmer, I have no any question about it. All I worried about was the context of the code.

-----------------------

Since I`ve stated my concerns several times above, now, I will just reply your questions here:
+ now `Settings.database_location = address` is in the outer indentation level and it is **reassigning** it in every database activity but it was written more elegant in original code
+ + `create_database` means == create a database
+ + `get_db` means == get a good database for the logged in user: **exists?** (_return the DB address & profile `id`_) : **not exists?** (_make a DB and return its address alongside profile `id` for the user_)

_I couldn't find a better name for it  rather than `get_db`._  

---------------------------------
  
Cheers 😁

**Pull request merged**